### PR TITLE
Create arraycmplen opcode to replace arraycmp + arrayCmpLen flag

### PR DESCRIPTION
**Pull request by Spencer-Comin**
See discussion at https://github.com/eclipse/omr/issues/6951
See also: matching OpenJ9 PR at https://github.com/eclipse-openj9/openj9/pull/17382

### CODE REVIEW

File: compiler/aarch64/codegen/OMRCodeGenerator.cpp
Comment (r30shah) [inline compiler/aarch64/codegen/OMRCodeGenerator.cpp:184]: Is there specific reason for using `TR_aarch64DisableArrayCmpLen` ?? i.e, why `aarch64` in the flag?

File: compiler/z/codegen/OMRTreeEvaluator.cpp
Comment (r30shah) [inline compiler/z/codegen/OMRTreeEvaluator.cpp:10298]: Checking out the [arraycmpSIMDHelper](https://github.com/eclipse/omr/blob/104d1f60f5a242078a88364fe5b78f54ef544db2/compiler/z/codegen/OMRTreeEvaluator.hpp#L1018-L1024), default last parameter is set to `false`, We would not need to set last parameter. 

Is there specific reason why you are explicitly setting it to `false` ? Or maybe I am missing something as I do see you added that in other places as well.

**Review by r30shah — CHANGES_REQUESTED** [review]

File: compiler/p/codegen/OMRTreeEvaluator.cpp
Comment (IBMJimmyk) [inline compiler/p/codegen/OMRTreeEvaluator.cpp:5629]: I think you accidentally removed the Power 10 implementation of arraycmp here.

Comment (Spencer-Comin) [reply compiler/p/codegen/OMRTreeEvaluator.cpp:5629]: I moved the if statement calling the Power 10 implementation to the [beginning of `inlineArrayCmp`](https://github.com/eclipse/omr/blob/104d1f60f5a242078a88364fe5b78f54ef544db2/compiler/p/codegen/OMRTreeEvaluator.cpp#L5446-L5448) so it wouldn't need to be duplicated between [`arraycmpEvaluator`](https://github.com/eclipse/omr/blob/104d1f60f5a242078a88364fe5b78f54ef544db2/compiler/p/codegen/OMRTreeEvaluator.cpp#L5632) and [`arraycmplenEvaluator`](https://github.com/eclipse/omr/blob/104d1f60f5a242078a88364fe5b78f54ef544db2/compiler/p/codegen/OMRTreeEvaluator.cpp#L5637)

File: compiler/aarch64/codegen/OMRCodeGenerator.cpp
Comment (Spencer-Comin) [reply compiler/aarch64/codegen/OMRCodeGenerator.cpp:184]: This is following the pattern for setting the `supportsArrayCmp` flag [a few lines earlier](https://github.com/eclipse/omr/blob/104d1f60f5a242078a88364fe5b78f54ef544db2/compiler/aarch64/codegen/OMRCodeGenerator.cpp#L179)

File: compiler/aarch64/codegen/OMRTreeEvaluator.cpp
Comment (Akira1Saitoh) [inline compiler/aarch64/codegen/OMRTreeEvaluator.cpp:6220]: When lengthReg is 64bit, `loadConstant32(cg, node, 0, lengthReg)` should be replaced with `loadConstant64(cg, node, 0, lengthReg)` for completeness. I am not able to put this comment to the exact location, but it is at line 6236.

**Comment by Spencer-Comin** [comment]
Back to draft PR while working out additional changes needed for Power codegen

**Comment by IBMJimmyk** [comment]
Here are the updated code changes I wrote for Power to handle the 32 bit and Power 10 cases:
https://github.com/IBMJimmyk/openj9-omr/commit/73f13640102097ed5347eb440551a514ed0a3581

File: compiler/z/codegen/OMRTreeEvaluator.cpp
Comment (r30shah) [inline compiler/z/codegen/OMRTreeEvaluator.cpp:11632]: Can you use `static_cast` here? Also just confirming, is `elemExps` `iconst` or `lconst`?

**Review by r30shah — APPROVED** [review]
@Spencer-Comin  Glanced through the Z codegen changes related to checking arrayCmpLen flag (As most of the evaluator is unchanged). Small nitpick which is not related to your change but something I found in existing code to change. Z changes LGTM

**Comment by r30shah** [comment]
@Akira1Saitoh @IBMJimmyk    Can we request your review for the codegen related changes ? 
@jdmpapin  @0xdaryl Can we request your review for the opcode related changes and x86 codegen changes ?

**Comment by Akira1Saitoh** [comment]
AArch64 codegen related changes look okay to me.

File: compiler/z/codegen/OMRTreeEvaluator.cpp
Comment (Spencer-Comin) [reply compiler/z/codegen/OMRTreeEvaluator.cpp:11632]: `elemsExpr` is the length child of the `arraycmp`. An `arraycmp` is generated in one place in OMR [1] and one place in OpenJ9 [2]. In OMR it generates an `imul` for the length, and that would be optimized to `iconst` to hit this codegen path. In OpenJ9 the length child is wrapped in an `i2l` if the platform is 64-bit, so it is optimized to `lconst` on 64-bit platform but `iconst` on 31-bit to hit this path.

In conclusion:
64-bit: `elemsExpr` could be `iconst` or `lconst` depending on where the `arraycmp` is generated
31-bit: `elemsExpr` is `iconst`

I could use `int64_t` instead of `int32_t` for `elems` here since `getIntegralValue` returns `int64_t` and the `MemCmpConstLenSignMacroOp` constructor (where `elems` is used later) takes an `int64_t`.

[1] https://github.com/Spencer-Comin/omr/blob/26a4bc82a7fb76a3aaed44dc923086b59957f4c0/compiler/optimizer/LoopReducer.cpp#L2162
[2] https://github.com/Spencer-Comin/openj9/blob/c68f163bea76f3385d6d786d503b1c2dc4756af0/runtime/compiler/optimizer/IdiomTransformations.cpp#L9603

**Comment by IBMJimmyk** [comment]
It looks good to me but I'm also the one who wrote most of the Power evaluator changes. @zl-wang Could you also try taking a look at `inlineArrayCmpP10` and `inlineArrayCmp` for Power? The vector implementation PR that I'm working on that is making changes to this area isn't quite ready yet. So at the moment these changes to the old implementation are needed for the opcode change work.

File: compiler/il/OMROpcodes.enum
Comment (r30shah) [inline compiler/il/OMROpcodes.enum:7210]: Looking at the other places where this property is used, does it apply to arrayCmpLen?

File: compiler/il/OMROpcodes.enum
Comment (r30shah) [inline compiler/il/OMROpcodes.enum:7220]: Now as arrayCmp and arrayCmpLen is different, this should be highlighted or atleast clarified in the description.

**Review by r30shah — CHANGES_REQUESTED** [review]
Gave another review for this change, minor nitpicks in the common code I can spot. As said before, Z code looks good to me.

**Review by jdmpapin — COMMENTED** [review]
Based on https://github.com/eclipse/omr/issues/6951#issuecomment-1540607730 the sequence generated on x86 looks good to me

It would probably also be good to test `arraycmplen` with large arrays (> 4GiB) and large difference offsets in ArrayTest.cpp, though I'm not sure how well CI would like that

File: compiler/p/codegen/OMRTreeEvaluator.cpp
Comment (jdmpapin) [reply compiler/p/codegen/OMRTreeEvaluator.cpp]: I see, this is fixed in the second commit. Would you mind correcting the first commit instead? Better to avoid introducing a bug than to introduce it and then fix it

File: compiler/optimizer/OMRSimplifierHandlers.cpp
Comment (Spencer-Comin) [reply compiler/optimizer/OMRSimplifierHandlers.cpp:12466]: I'll add a check for the sign flag, but in a separate commit as it is out of the scope of the two commits already in this PR

**Comment by jdmpapin** [comment]
Mac failure is #6516

**Review by jdmpapin — APPROVED** [review]

**Comment by jdmpapin** [comment]
I'm taking the latest comments from @IBMJimmyk and @Akira1Saitoh as approvals

@r30shah, please approve if you have no further comments

@zl-wang, we still need your review as Jimmy requested

**Review by r30shah — APPROVED** [review]
I am good with all the changes.

**Comment by Spencer-Comin** [comment]
Reverting to draft while I look into some failing tests (see https://github.com/eclipse-openj9/openj9/pull/17382#issuecomment-1602649424)

**Comment by Spencer-Comin** [comment]
I did some tweaking to this PR and https://github.com/eclipse-openj9/openj9/pull/17382 to resolve failures in x86-32, it should be good to go now

**Comment by Spencer-Comin** [comment]
@r30shah

**Comment by r30shah** [comment]
@jdmpapin  Can I request to review and merge this change?

File: compiler/x/codegen/OMRTreeEvaluator.cpp
Comment (jdmpapin) [inline compiler/x/codegen/OMRTreeEvaluator.cpp:1304]: I doubt that `MOV8RegReg` works on 32-bit. Please use `longClobberEvaluate()`. On 64-bit it will do this, but on 32-bit it will use a pair of `MOV4RegReg` instructions instead

Comment (Spencer-Comin) [reply compiler/x/codegen/OMRTreeEvaluator.cpp]: `longClobberEvaluate` is the functionality I wanted, thanks for bringing it up

**Review by jdmpapin — APPROVED** [review]

**Comment by jdmpapin** [comment]
I'm not sure why the Mac build was cancelled. I don't seem to be able to restart it either. @Spencer-Comin, would you mind amending your commit without making any changes? If the commit date changes, it will get a different SHA and the automatic checks should restart

**Comment by jdmpapin** [comment]
Jenkins build all

**Comment by jdmpapin** [comment]
Jenkins build all

**Comment by Spencer-Comin** [comment]
Converting to draft while I make `arraycmp(len)` enablement more uniform across platforms, see https://github.com/eclipse-openj9/openj9/pull/16662#discussion_r1317469710

**Comment by jdmpapin** [comment]
I've been trying to get this merged for a while, but I've really been making a push with the PR builds in the last 24 hours or so - see the all activity on eclipse-openj9/openj9#17382. Would it be OK to make these further changes in a separate PR once this one is merged?

**Pull request merged**
